data_validation/schema_validation.py

```diff
@@ -216,40 +228,99 @@ def schema_validation_matching(
+    ]
+    invalid_pairs = [_ for _ in data_type_pairs if len(_) != 2]
+    if invalid_pairs:
+        raise exceptions.SchemaValidationException(
+            f"Invalid data type pairs: {invalid_pairs}"
+        )
+    return data_type_pairs
+
+
+def expand_precision_range(s: str) -> list[str]:
+    """Expand an integer range (e.g. "0-3") to a list (e.g. ["0", "1", "2", "3"])."""
+    m_range = DECIMAL_PRECISION_SCALE_RANGE_PATTERN.match(s)
+    if not m_range:
+        return [s]
+    try:
+        p_lower = int(m_range.group(1))
+        p_upper = int(m_range.group(2))
+        if p_lower >= p_upper:
+            raise exceptions.SchemaValidationException(
+                f"Invalid allow list data type precision/scale: Lower value {p_lower} >= upper value {p_upper}"
+            )
+        return [str(_) for _ in range(p_lower, p_upper + 1)]
+    except ValueError as e:
+        raise exceptions.SchemaValidationException(
+            f"Invalid allow list data type precision/scale: {s}"
+        ) from e
 
 
-def get_datatype_name(st):
-    chars = []
-    for i in range(len(st)):
-        if ord(st[i].lower()) >= 97 and ord(st[i].lower()) <= 122:
-            chars.append(st[i].lower())
-    out = "".join(chars)
-    if out == "":
-        return -1
-    return out
+def expand_precision_or_scale_range(data_type: str) -> list[str]:
+    """Take a data type and example any precision/scale range.
+
+    For example "decimal(1-3,0)" becomes:
+      ["decimal(1,0)", "decimal(2,0)", "decimal(3,0)"]"""
+
+    m = DECIMAL_PRECISION_SCALE_PATTERN.match(data_type.replace(" ", ""))
+    if not m:
+        return [data_type]
+
+    if len(m.groups()) != 3:
+        raise exceptions.SchemaValidationException(
+            f"Badly formatted data type: {data_type}"
+        )
+
+    type_name, p, s = m.groups()
+    p_list = expand_precision_range(p)
+    if s:
+        s_list = expand_precision_range(s)
+        return_list = [
+            f"{type_name}({p},{s})" for p, s in itertools.product(p_list, s_list)
+        ]
+    else:
+        return_list = [f"{type_name}({_})" for _ in p_list]
+    return return_list
+
+
+def parse_allow_list(st: str) -> dict:
+    """Convert allow-list data type pairs into a dictionary like {key[value1, value2, etc], }"""
+
+    def expand_allow_list_ranges(data_type_pairs: list[tuple]) -> list[tuple]:
+        expanded_pairs = []
+        for dt1, dt2 in data_type_pairs:
+            dt1_list = expand_precision_or_scale_range(dt1)
+            dt2_list = expand_precision_or_scale_range(dt2)
+            expanded_pairs.extend(
+                [(_[0], _[1]) for _ in itertools.product(dt1_list, dt2_list)]
+            )
+        return expanded_pairs
+
+    def convert_pairs_to_dict(expanded_pairs: list) -> dict:
+        return_pairs = {}
+        for dt1, dt2 in expanded_pairs:
+            if dt1 in return_pairs:
+                return_pairs[dt1].append(dt2)
```

Comment:
shouldn't this be `dt[1]=dt[2]`?

Reply:
if not, can you provide a description of what this function is doing?

Reply:
Did that satisfy the query?
Or do you think my code should use a list called `dt` instead of two variables `dt1` and `dt2`?

Reply:
Thanks, docstring helps